### PR TITLE
Add a woocommerce_check_cart_item_stock_ignore_ids filter to allow bypass of check_cart_item_stock()

### LINKED ISSUE #26944: Add filter/way of overriding cart items got in check_cart_item_stock()
[check_cart_item_stock()](https://github.com/woocommerce/woocommerce/blob/7fa34c6770c4f55d2071e48f69f386ed15504553/includes/class-wc-cart.php#L757) is used to deal with cart items that are no longer available.

In some custom development scenarios you may want to omit specific cart items from this check. It uses `$this->get_cart()` as a means of getting the cart items and therefore we can't see a way to manipulate this other than using [wp_debug_backtrace_summary](https://developer.wordpress.org/reference/functions/wp_debug_backtrace_summary/) to find out if `do_action('woocommerce_check_cart_items')` is a caller (which eventually fires the check_cart_item_stock function).

### PR DESCRIPTION
**Pull request by kevinruscoe**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

check_cart_item_stock() is used to deal with cart items that are no longer available.

In some custom development scenarios you may want to omit specific cart items from this check. It uses $this->get_cart() as a means of getting the cart items and therefore we can't see a way to manipulate this other than using wp_debug_backtrace_summary to find out if do_action('woocommerce_check_cart_items') is a caller (which eventually fires the check_cart_item_stock function).

This adds a `woocommerce_check_cart_item_stock_ignore_ids` filter that allows you to pass an array of IDs for the check to ignore.

Closes #26944 .

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Add a `woocommerce_check_cart_item_stock_ignore_ids` filter so you can ignore certain products when `check_cart_item_stock()` is ran.

### CODE REVIEW

**Comment by ninetyninew** [comment]
Hey, just to let you know I am going to do further work on this to also take advantage of the cart item key too, this way you can remove products from this check based on the cart item key which in turn means you can remove products from the checked based on any specific meta the cart item may have in relation to this - in some scenarios the product id would not be enough.

**Comment by kevinruscoe** [comment]
@ninetyninew Sounds good, I'd suggest to copy the `$cart_item_ids_to_ignore = apply_filters( 'woocommerce_check_cart_item_stock_ignore_ids', array() );` line, to something like `$cart_line_item_ids_to_ignore = apply_filters( 'woocommerce_cart_line_item_ids_to_ignore', array() );`, then merge the 2 arras together, and use that final array to check against in the loop below.

Whilst you're at it, it might be worth coming up with some better filter names. Tbh, I don't think they're clear enough what they do and they're pretty generic.

**Comment by ninetyninew** [comment]
This should do it https://github.com/ninetyninew/woocommerce/commit/386a491a6aff5e10600959edb0a4e9281678ca6e

What do you think? This means you can ignore specific product ids where needed and also ignore by cart item key.

The cart item key is important as you may have items added to the cart containing specific meta which means check_cart_item_stock should not be used.

I can either do this as a new pull request or you update yours?

**Comment by rodrigoprimo** [comment]
Thanks for your contribution, @kevinruscoe! Instead of creating a filter to populate a variable with a list of IDs to ignore, what if we created a filter to modify whatever is returned by `$this->get_cart()`? I believe this is a more flexible and less specific solution. Would this work for your use case?

This is what I have in mind (not super happy about the filter name):

```php
$cart_items = apply_filters( 'woocommerce_check_cart_item_stock_items', $this->get_cart() );

foreach ( $cart_items as $cart_item_key => $values ) {
   ...
}
```

File: includes/class-wc-cart.php [hunk 1]
```diff
@@ -758,9 +758,14 @@ public function check_cart_item_stock() {
 		$error                    = new WP_Error();
 		$product_qty_in_cart      = $this->get_cart_item_quantities();
 		$current_session_order_id = isset( WC()->session->order_awaiting_payment ) ? absint( WC()->session->order_awaiting_payment ) : 0;
+		$cart_item_ids_to_ignore = apply_filters( 'woocommerce_check_cart_item_stock_ignore_ids', array() );
```
Comment (claudiosanches) [inline includes/class-wc-cart.php:761]: This is very custom, and don't have any use on core by default.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Thanks for your help, but there's other alternatives to do what you intent to do.

Like it's possible to use the `woocommerce_get_cart_contents` checking if `woocommerce_check_cart_items` got triggered with `has_filter()`.

**Comment by ninetyninew** [comment]
@claudiosanches Thanks for the info on `woocommerce_get_cart_contents` checking if `woocommerce_check_cart_items` got triggered with `has_filter()`. Although I'm struggling to figure out how this would help with stopping the stock check for a specific cart item, I tested this:

`
add_filter( 'woocommerce_get_cart_contents', 'test123' );
public function test123( $cart_contents ) {

	if( has_filter( 'woocommerce_check_cart_items' ) ) {

		foreach( $cart_contents as $k => $v ) {

			if( isset( $v['some_meta_key'] ) ) { // some_meta_key would be a key we define to stop this cart item being checked

				unset( $cart_contents[$k] );

			}

		}
	
	}

	return $cart_contents;

}
`

But this would still remove the cart item when loading the cart. So while I get why this will not be considered for core, I can't see any way we could use anything WooCommerce includes in core to disable the stock check for specific cart items, which while we are trying to do this for a specific reason (rental products) I suspect there may be other scenarios where a developer building something custom would want to bypass the cart stock check on a specific item - so I feel like `woocommerce_check_cart_items` does limit the extensibility possible. Maybe you were thinking of another means of using those filters I am unaware of? Any pointers here would be really appreciated.

**Comment by ninetyninew** [comment]
Just incase anyone is looking for similar info on this kind of functionality, I saw this recently, it might help - not yet tested this myself: https://wordpress.stackexchange.com/questions/378040/wooocommerce-disable-check-cart-item-stock-if-cart-item-has-specific-meta

**Pull request closed without merging**